File: Apps/Api/lib/appserver-bridge.js
const { spawn } = require("child_process");
const path = require("path");

const DEFAULT_PYTHON_RUNTIME_ROOT = path.resolve(__dirname, "..", "..", "..", "Infrastructure", "AgentRuntime");

function createAppServerBridge({
  pythonRuntimeRoot = process.env.PYTHON_RUNTIME_ROOT || DEFAULT_PYTHON_RUNTIME_ROOT,
  python = process.env.PYTHON || "python",
} = {}) {
  const bridgePath = path.join(__dirname, "appserver_bridge.py");

  async function runTurnWithInputs({ workspaceRoot, threadId, inputs, skillPath, timeoutSeconds = 180 }) {
    const payload = {
      operation: "runTurnWithInputs",
      pythonRuntimeRoot,
      workspaceRoot,
      threadId,
      inputs,
      skillPath,
      timeoutSeconds,
      transportUrl: process.env.CODEX_APP_SERVER_WS_URL || "ws://127.0.0.1:8146",
    };
    const result = await runPythonJson({ python, script: bridgePath, payload, timeoutMs: (timeoutSeconds + 30) * 1000 });
    if (!result?.ok) {
      const error = new Error(result?.message || "AppServer turn failed");
      error.code = result?.error || "appserver_turn_failed";
      error.debugPayload = result;
      throw error;
    }
    return result;
  }

  async function startTurnWithInputs({ workspaceRoot, threadId, inputs, skillPath, timeoutSeconds = 180 }) {
    const payload = {
      operation: "startTurnWithInputs",
      pythonRuntimeRoot,
      workspaceRoot,
      threadId,
      inputs,
      skillPath,
      timeoutSeconds,
      transportUrl: process.env.CODEX_APP_SERVER_WS_URL || "ws://127.0.0.1:8146",
    };
    const result = await runPythonJson({ python, script: bridgePath, payload, timeoutMs: 45000 });
    if (!result?.ok) throw appServerError(result, "appserver_turn_start_failed");
    return result;
  }

  async function collectTurnResult({ workspaceRoot, threadId, turnId, timeoutSeconds = 180 }) {
    const payload = {
      operation: "collectTurnResult",
      pythonRuntimeRoot,
      workspaceRoot,
      threadId,
      turnId,
      timeoutSeconds,
      transportUrl: process.env.CODEX_APP_SERVER_WS_URL || "ws://127.0.0.1:8146",
    };
    const result = await runPythonJson({ python, script: bridgePath, payload, timeoutMs: 45000 });
    if (!result?.ok && !isNonTerminalTurnStatus(result?.status)) throw appServerError(result, "appserver_turn_collect_failed");
    return result;
  }

  async function readThread({ workspaceRoot, threadId, timeoutSeconds = 60 }) {
    const payload = {
      operation: "readThread",
      pythonRuntimeRoot,
      workspaceRoot,
      threadId,
      timeoutSeconds,
      transportUrl: process.env.CODEX_APP_SERVER_WS_URL || "ws://127.0.0.1:8146",
    };
    const result = await runPythonJson({ python, script: bridgePath, payload, timeoutMs: 45000 });
    if (!result?.ok) throw appServerError(result, "appserver_thread_read_failed");
    return result;
  }

  return { pythonRuntimeRoot, runTurnWithInputs, startTurnWithInputs, collectTurnResult, readThread };
}

function appServerError(result, fallbackCode) {
  const error = new Error(result?.message || "AppServer turn failed");
  error.code = result?.error || fallbackCode;
  error.debugPayload = result;
  return error;
}

function runPythonJson({ python, script, payload, timeoutMs }) {
  return new Promise((resolve, reject) => {
    const child = spawn(python, [script], {
      stdio: ["pipe", "pipe", "pipe"],
      env: { ...process.env, PYTHONIOENCODING: process.env.PYTHONIOENCODING || "utf-8" },
    });
    let stdout = "";
    let stderr = "";
    const timer = setTimeout(() => {
      child.kill();
      const error = new Error("AppServer bridge timed out");
      error.code = "appserver_bridge_timeout";
      error.debugPayload = { stderr: stderr.slice(-2000) };
      reject(error);
    }, timeoutMs);
    child.stdout.on("data", (chunk) => {
      stdout += chunk.toString("utf8");
    });
    child.stderr.on("data", (chunk) => {
      stderr += chunk.toString("utf8");
    });
    child.on("error", (error) => {
      clearTimeout(timer);
      error.debugPayload = { stderr: stderr.slice(-2000) };
      reject(error);
    });
    child.on("close", (code) => {
      clearTimeout(timer);
      if (code !== 0) {
        const parsed = tryParseJson(stdout);
        const structured = parsed && typeof parsed === "object" ? parsed : null;
        const error = new Error(structured?.message || `AppServer bridge exited with ${code}`);
        error.code = structured?.error || "appserver_bridge_failed";
        error.debugPayload = {
          stderr: stderr.slice(-2000),
          stdout: stdout.slice(-2000),
          structured: structured ?? null,
        };
        reject(error);
        return;
      }
      try {
        resolve(JSON.parse(stdout));
      } catch (error) {
        error.debugPayload = { stdout: stdout.slice(-2000), stderr: stderr.slice(-2000) };
        reject(error);
      }
    });
    child.stdin.end(JSON.stringify(payload));
  });
}

function tryParseJson(text) {
  try {
    return JSON.parse(text);
  } catch {
    return null;
  }
}

module.exports = { DEFAULT_PYTHON_RUNTIME_ROOT, createAppServerBridge, isNonTerminalTurnStatus };

function isNonTerminalTurnStatus(status) {
  return ["created", "pending", "queued", "submitted", "running", "inprogress", "in_progress"].includes(String(status ?? "").trim().toLowerCase());
}

File: Apps/Api/lib/appserver_bridge.py
from __future__ import annotations

import json
import sys
from pathlib import Path


def sanitize_for_appserver_text(value):
    if isinstance(value, str):
        return "".join(ch for ch in value if not 0xD800 <= ord(ch) <= 0xDFFF)
    if isinstance(value, list):
        return [sanitize_for_appserver_text(item) for item in value]
    if isinstance(value, dict):
        return {key: sanitize_for_appserver_text(item) for key, item in value.items()}
    return value


def main() -> int:
    payload = sanitize_for_appserver_text(json.loads(sys.stdin.read() or "{}"))
    cep_root = Path(payload["cepRoot"]).resolve()
    sys.path.insert(0, str(cep_root))

    from ae_workspace_core.appserver.client import AppServerSessionClient  # noqa: WPS433

    client = AppServerSessionClient(
        payload["workspaceRoot"],
        transport_mode="ws",
        transport_url=payload.get("transportUrl") or "ws://127.0.0.1:8146",
        request_timeout_seconds=30.0,
        turn_timeout_seconds=float(payload.get("timeoutSeconds") or 180),
        init_timeout_seconds=90.0,
    )
    client.start()
    try:
        turn_id = client.start_turn_with_inputs(
            str(payload["threadId"]),
            list(payload.get("inputs") or []),
            skill_path=payload.get("skillPath") or None,
            cwd=payload["workspaceRoot"],
        )
        status = client.wait_turn_completed(
            str(payload["threadId"]),
            turn_id,
            timeout_seconds=float(payload.get("timeoutSeconds") or 180),
        )
        result = client.collect_turn_result(
            thread_id=str(payload["threadId"]),
            turn_id=turn_id,
            return_thread_state=False,
        )
        sys.stdout.write(
            json.dumps(
                {
                    "ok": status == "completed" and result.status == "completed",
                    "threadId": result.thread_id,
                    "turnId": result.turn_id,
                    "status": result.status,
                    "finalMessage": result.final_message,
                },
                ensure_ascii=False,
            )
        )
        return 0
    finally:
        client.close()


if __name__ == "__main__":
    raise SystemExit(main())
